File: Example-iOS-ObjC/ViewController.m
//
//  ViewController.m
//  Example-iOS-ObjC
//
//  Created by Darren Jones on 02/01/2020.
//  Copyright © 2020 Darren Jones. All rights reserved.
//

#import "ViewController.h"
#import <DJLogging/DJLogging.h>
@import MessageUI;
#import "DJLogTypeUI.h"
#import "DJLogTypeComms.h"

@interface ViewController () <MFMailComposeViewControllerDelegate>

@end

@implementation ViewController

- (void)viewDidLoad
{
    LogMethodCallWithType(DJLogTypeUI.shared)
    
    [super viewDidLoad];
    
    [self makeAWebRequest];
}

- (void)makeAWebRequest
{
    NSUUID *uuid = [NSUUID UUID];
    LogMethodCallWithUUIDAndType(uuid, DJLogTypeComms.shared)
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://venderbase.com"]];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    NSURLSessionDataTask *task = [session dataTaskWithRequest:request
                                            completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        LogRequestResponseWithUUIDAndType(uuid, DJLogTypeComms.shared)
    }];
    [task resume];
}

- (IBAction)emailSupportButtonPressed:(id)sender
{
    LogMethodCallWithType(DJLogTypeUI.shared)
    
    if ([MFMailComposeViewController canSendMail])
    {
        MFMailComposeViewController *mailer = [[MFMailComposeViewController alloc] init];
        [mailer setMailComposeDelegate:self];
        
        NSString *appName = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"];
        [mailer setSubject:[NSString stringWithFormat:@"%@ Support Request (iOS)", appName]];
        NSArray *toRecipients = [NSArray arrayWithObject:@"add.support.email@here.com"];
        [mailer setToRecipients:toRecipients];
        
        // Fetch the log
        NSData *logData = [LogManager htmlData];
        
        // Attach the HTML log
        [mailer addAttachmentData:logData mimeType:@"text/html" fileName:@"log.html"];
        
        // Present new email composer
        [self presentViewController:mailer animated:YES completion:nil];
    }
    else
    {
        // If we cannot send mail (eg. Simulator) then dump the log to NSLog
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Error", nil)
                                                                                 message:NSLocalizedString(@"Your device doesn't support sending email", nil)
                                                                          preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
                                                               style:UIAlertActionStyleCancel
                                                             handler:nil];
        [alertController addAction:cancelAction];
        
        [self presentViewController:alertController animated:YES completion:nil];
        
        [LogManager printLogs];
        
        NSData *logData = [LogManager htmlData];
        NSURL *temporaryDirectoryURL = [NSURL fileURLWithPath:NSTemporaryDirectory() isDirectory:YES];
        NSURL *fileURL = [temporaryDirectoryURL URLByAppendingPathComponent:@"logs.html"];
        NSLog(@"%@", fileURL);
        
        [logData writeToURL:fileURL atomically:YES];
    }
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    switch (result)
    {
        case MFMailComposeResultCancelled:
            NSLog(@"Mail cancelled: you cancelled the operation and no email message was queued.");
            break;
        case MFMailComposeResultSaved:
            NSLog(@"Mail saved: you saved the email message in the drafts folder.");
            break;
        case MFMailComposeResultSent:
            NSLog(@"Mail send: the email message is queued in the outbox. It is ready to send.");
            break;
        case MFMailComposeResultFailed:
            NSLog(@"Mail failed: the email message was not saved or queued, possibly due to an error.");
            break;
        default:
            NSLog(@"Mail not sent.");
            break;
    }
    // Remove the mail view
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Example-iOS-Swift/ViewController.swift
//
//  ViewController.swift
//  Example-iOS
//
//  Created by Darren Jones on 01/01/2020.
//  Copyright © 2020 Darren Jones. All rights reserved.
//

import UIKit
import MessageUI
import DJLogging

class ViewController: UIViewController, MFMailComposeViewControllerDelegate {
    
    override func viewDidLoad() {
        
        LogMethodCall(type: .ui)
        
        super.viewDidLoad()
        
        makeAWebRequest()
    }
    
    func makeAWebRequest() {
        let uuid = UUID()
        LogMethodCall(uuid, type: .comms)
        
        let request = URLRequest(url: URL(string: "https://venderbase.com")!)
        let session = URLSession(configuration: .default)
        let task = session.dataTask(with: request) { (data, response, error) in
            
            LogRequestResponse(uuid: uuid, response: response, data: data, error: error, type: .comms)
        }
        task.resume()
    }
    
    @IBAction func emailSupportButtonPressed(_ sender: Any) {
        
        LogMethodCall(type: .ui)
        
        if MFMailComposeViewController.canSendMail()
        {
            let mailComposeViewController = configuredMailComposeViewController()
            
            // Present new email composer
            self.present(mailComposeViewController, animated: true, completion: nil)
        }
        else
        {
            // If we cannot send mail (eg. Simulator) then dump the log to Debug
            let alertController = UIAlertController(title: NSLocalizedString("Error", comment: ""),
                                                    message: NSLocalizedString("Your device doesn't support sending email", comment: ""),
                                                    preferredStyle: .alert)
            
            let cancelAction = UIAlertAction(title: NSLocalizedString("OK", comment: ""),
                                             style: .cancel,
                                             handler: nil)
            
            alertController.addAction(cancelAction)
            
            self.present(alertController, animated: true, completion: nil)
            
            LogManager.printLogs()
        }
    }
    
    func configuredMailComposeViewController() -> MFMailComposeViewController {
        let mailer = MFMailComposeViewController()
        mailer.mailComposeDelegate = self;
        
        let appName = Bundle.main.object(forInfoDictionaryKey: "CFBundleName")
        mailer.setSubject("\(appName ?? "") Support Request (iOS)")
        mailer.setToRecipients(["add.support.email@here.com"])
        
        // Convert log to HTML and attach file to email
        let htmlData = LogManager.htmlData() ?? Data()
        mailer.addAttachmentData(htmlData, mimeType: "text/html", fileName: "log.html")
        
        return mailer
    }
    
    // MARK: - MFMailComposeViewControllerDelegate
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        
        switch result {
        case .cancelled:
            print("Mail cancelled: you cancelled the operation and no email message was queued.")
            
        case .saved:
            print("Mail saved: you saved the email message in the drafts folder.")
            
        case .sent:
            print("Mail send: the email message is queued in the outbox. It is ready to send.")
            
        case .failed:
            print("Mail failed: the email message was not saved or queued, possibly due to an error.")
            
        default:
            print("Mail not sent.")
        }
        
        // Remove the mail view
        self.dismiss(animated: true, completion: nil)
    }
    
}

